php/settings/class-setting.php

```diff
@@ -0,0 +1,678 @@
+		}
+
+		return ! is_null( $setting ) ? $setting : $this->get_setting( $slug ); // return a dynamic setting.
+	}
+
+	/**
+	 * Recursivly find a setting.
+	 *
+	 * @param string $slug The setting slug to get.
+	 *
+	 * @return Setting|null
+	 */
+	protected function find_setting_recursively( $slug ) {
+		$found = null;
+		// loop through settings to find it.
+		foreach ( $this->get_settings() as $sub_setting ) {
+			if ( $sub_setting->has_settings() ) {
+				if ( $sub_setting->has_setting( $slug ) ) {
+					$found = $sub_setting->get_setting( $slug );
+					break;
+				}
+				$found = $sub_setting->find_setting_recursively( $slug );
+				if ( ! is_null( $found ) ) {
+					break;
+				}
+			}
+		}
+
+		return $found;
+	}
+
+	/**
+	 * Register a setting.
+	 *
+	 * @param array $params The setting params.
+	 */
+	public function setup_setting( array $params ) {
+		$dynamic_params = array_filter( $params, array( $this, 'is_setting_param' ), ARRAY_FILTER_USE_KEY );
+		foreach ( $params as $param => $value ) {
+			if ( $this->is_setting_param( $param ) ) {
+				continue;
+			}
+			// Set params.
+			$this->set_param( $param, $value );
+		}
+
+		// Register dynamics.
+		$this->register_dynamic_settings( $dynamic_params );
+	}
+
+	/**
+	 * Register the setting with WordPress.
+	 */
+	protected function register_setting() {
+		$option_group = $this->get_option_name();
+		if ( ! is_null( $option_group ) && ! $this->has_settings() && ! $this->get_parent()->has_param( 'setting_registered' ) ) {
+			$args = array(
+				'type'              => 'array',
+				'description'       => $this->get_param( 'description' ),
+				'sanitize_callback' => array( $this, 'prepare_sanitizer' ),
+				'show_in_rest'      => false,
+			);
+			register_setting( $option_group, $option_group, $args );
+			add_filter( 'pre_update_site_option_' . $option_group, array( $this, 'set_notices' ), 10, 3 );
+			add_filter( 'pre_update_option_' . $option_group, array( $this, 'set_notices' ), 10, 3 );
+			$this->get_parent()->set_param( 'setting_registered', true );
+		}
+	}
+
+	/**
+	 * Prepare the setting option group to be sanitized by each component.
+	 *
+	 * @param array $data Array of values to sanitize.
+	 *
+	 * @return array
+	 */
+	public function prepare_sanitizer( $data ) {
+		$slug = $this->get_slug();
+		if ( isset( $data[ $slug ] ) ) {
+			$data[ $slug ] = $this->get_component()->sanitize_value( $data[ $slug ] );
+		}
+
+		return $data;
+	}
+
+	/**
+	 * Set notices on successful update of settings.
+	 *
+	 * @param mixed  $value        The new Value.
+	 * @param mixed  $old_value    The old value.
+	 * @param string $setting_slug The setting key.
+	 *
+	 * @return mixed
+	 */
+	public function set_notices( $value, $old_value, $setting_slug ) {
+
+		if ( $value !== $old_value ) {
+			if ( is_wp_error( $value ) ) {
+				add_settings_error( $setting_slug, 'setting_notice', $value->get_error_message(), 'error' );
+
+				return $old_value;
+			} else {
```

Comment:
@pereirinha - the mistake here is that I'm returning where I don't need to.
I prefer returning at the bottom, not mid method if it can be avoided. Makes testing paths easier without having multiple exit points.
I'll address it.

Reply:
That makes sense, and I agree that a single return is the best in this case.